docs/core/testing/unit-testing-best-practices.md

```diff
@@ -0,0 +1,361 @@
+---
+title: Best practices for writing unit tests
+description: Learn best practices for writing unit tests that drive code quality and resilience
+author: jpreese
+ms.author: wiwagn
+ms.date: 07/28/2018
+---
+
+# Unit testing best practices
+
+By [John Reese](http://reesespieces.io) with special thanks to [Roy Osherove](http://osherove.com/)
+
+There are numerous benefits to writing unit tests; they help with regression, provide documentation, and facilitate good design. However, hard to read and brittle unit tests can wreak havoc on your code base.
+
+In this guide, you'll learn some best practices when writing unit tests to keep your tests resilient and easy to understand.
+
+## Why unit test?
+
+### Less time performing functional tests
+Functional tests are expensive. They typically involve opening up the application and performing a series of steps that you (or someone else), must follow in order to validate the expected behavior. These steps may not always be known to the tester, which means they will have to reach out to someone more knowledgeable in the area in order to carry out the test. Testing itself could take seconds for trivial changes, or minutes for larger changes. Lastly, this process must be repeated for every change that you make in the system.
+
+Unit tests, on the other hand, take milliseconds, can be run at the press of a button and do not necessarily require any knowledge of the system at large. Whether or not the test passes or fails is up to the test runner, not the individual.
+
+### Protection against regression
+Regression defects are defects that are introduced when a change is made to the application. It is common for testers to not only test their new feature but also features that existed beforehand in order to verify that previously implemented features still function as expected.
+
+With unit testing, it's possible to rerun your entire suite of tests after every build or even after you change a line of code. Giving you confidence that your new code does not break existing functionality.
+
+### Executable documentation
+It may not always be obvious what a particular method does or how it behaves given a certain input. You may ask yourself: How does this method behave if I pass it a blank string? Null?
```

Comment:
Having read this enough times, I have to leave this one last comment.

Every time I read this, I found myself thinking. "You may ask yourself, How do I work this?"  😈